src/library_exceptions.js

```diff
@@ -398,29 +398,32 @@ var LibraryExceptions = {
     return Module['asm']['__cpp_exception'];
   },
 
-  $getCppExceptionThrownValue__deps: ['$getCppExceptionTag'],
+  $getCppExceptionThrownValue__deps: ['$getCppExceptionTag', '__thrown_object_from_unwind_exception'],
   $getCppExceptionThrownValue: function(ex) {
```

Comment:
Should we call this function `getCppExceptionThrownObject`?  Or is thrown value a thing that is different?

Reply:
Yeah these are confusing. In library_exceptions.js, I tried to use 'object' for `WebAssembly.Exception` object thrown, and 'value' for the user-thrown value. So this takes an object and returns the user-thrown value. This got more confusing by the param name for this function is `ex` while other functions' param name is `obj`, when they denote the same thing.

At the same time libc++abi uses 'object' to mean the user-thrown value, which adds to the confusion.. I guess we should make these terms clearer.

How about this?
- object: the thrown value. So we rename this function to `getCppExceptionThrownObject` as you suggested.
- value: let's not use this term at all
- ex/exception: `WebAssembly.Exception`. So we rename current `obj` params in other functions, which meant `WebAssembly.Exceptions`, to `ex`.

If you think this sound fine, I'll make this change as a followup PR.

Reply:
Maybe "JSObject" or "JSException"?

Reply:
What is the thing you want to represent with "JSObject" or "JSException"? `WebAssembly.Exception` or the user-thrown value(or object)? I guess the latter? In that case it's confusing in another way because this is a Wasm exception and that sounds like this is a JS exception thrown from JS, a foreign exception..

Reply:
`WebAssembly.Exception` is what I propose to refer to as JS object.   The thrown object lives in wasm memory, as does the exception object, right?  They are both allocated with malloc.

Reply:
This is how I've been thinking about it:   JS Object (`WebAssembly.Exception`) -> refers to a C++ `Exception` class object allocated by libc++ -> refers to a user object that that user threw.  

JSExceptionObject -> C++ExceptionObject -> UserObject/ThrownValue

Reply:
Sorry I'm not sure if I understand. `WebAssembly.Exception` is not what the user threw, and this is the source of all these term differences. 

> JSExceptionObject -> C++ExceptionObject -> UserObject/ThrownValue

But this sounds like these three are the same? Is that what you mean?

Reply:
The user throws a C++ object of their choosing (the thrown object), libc++ then wraps then into an Exception object (still native pointer to memory), and then this can be caught on the outside as a JavaScript exception object.

Is that about right?

They do all kind of all refer to same thing, but they are distinct right?     And in particular that is no way to pass the JSObject into native code (since its not a number like the other two).

Reply:
The user throws a C++ object (native pointer), and what libc++abi throws is a pointer to `_Unwind_Exception`, which still a native pointer but an address preceding the user-thrown project. In this layout, the start address of the third block (thrown object) is what an user originally throws, and the start address of `_Unwind_Exception` is what libc++abi ends up throwing.
https://github.com/emscripten-core/emscripten/blob/24f765dc3545e89f232c9f8503f6426055271628/system/lib/libcxxabi/src/cxa_exception.cpp#L26-L28

`WebAssembly.Exception` is something different. This has nothing to do with libc++abi. libc++abi calls libunwind to do the actual low level throwing stuff from the VM: https://github.com/emscripten-core/emscripten/blob/24f765dc3545e89f232c9f8503f6426055271628/system/lib/libcxxabi/src/cxa_exception.cpp#L284

And our libunwind port calls `__builtin_wasm_throw`, which is lowered down to a wasm `throw` instruction: https://github.com/emscripten-core/emscripten/blob/24f765dc3545e89f232c9f8503f6426055271628/system/lib/libunwind/src/Unwind-wasm.c#L50-L56

And within this Wasm native `throw` instruction, it wraps the value thrown by libc++abi (a pointer to `_Unwind_Exception`) into a `WebAssembly.Exception` object, and throws that object.

Nevermind `_Unwind_Exception` thing here, which I don't want to expose. What we are confused about is the use of the term 'object'. Does it denote the user-thrown native pointer, or `WebAssembly.Exception` object?

I still think `JSException` is confusing because this sounds like a foreign (non-wasm) exception..

Reply:
I see, so `WebAssembly.Exception`, despite being a JavaScript object, starts life on the native/wasm side, so it corresponds to a wasm object?   On the wasm side its a tag/i32 pair or something like that?

However in this code it really is a JS heap object, right?    This function take a JS heap object and returns a pointer/number, right?

Reply:
For example a longer name for this function might be something like `getCppThrownObjectFromJSCaughtObject`?

Reply:
Not sure `JSCaughtObject` is clearer to someone who doesn't know this context. How about just `getCppThrownObjectFromWebassemblyException`, with comments clarifying what this means? This name is long but I guess at least not misleading..

Reply:
Sure, sounds reasonable.   I think if was a user I would be looking for something do with the thing that I caught.. so I guess `JSCaughtObject` and `WebassemblyException` both describe that thing.

Reply:
I'll explain in the comment that this is `WebAssembly.Exception` object that Wasm EH JS API provides.

Reply:
Anyway, will do this in a followup. Thanks for the discussion!